File: src/algorithms/sort/c/quickSort.c
#include <stdio.h>

static void swap(int *values, int i, int j){
    if (i != j){
        int temp = values[i];
        values[i] = values[j];
        values[j] = temp;
    }
}

static void partition1(int *values, int start, int end){
    if (end <= start){
        return;
    }

    int i = start;

    for (int j = start; j < end; j++){
        if (values[j] < values[end]){
            swap(values, i, j);
            i += 1;
        }
    }

    swap(values, i, end);
    partition1(values, start, i-1);
    partition1(values, i+1, end);
}

static void partition2(int *values, int start, int end){
    if (end <= start) {
        return;
    }

    int ref = values[start];
    int left = start + 1;
    int right = end;

    while (left <= right) {
        if (values[left] > ref && values[right] < ref) {
            swap(values, left, right);
        }

        if (values[left] <= ref) {
            left += 1;
        }

        if (values[right] >= ref) {
            right -= 1;
        }
    }

    int pivotIndex = right;
    swap(values, pivotIndex, start);

    partition2(values, start, pivotIndex - 1);
    partition2(values, pivotIndex + 1, end);
}


int quickSort1(int *arr, int arrLength){
    if (arrLength <= 0){
        printf("Array length must be greater than 0");
        return -1;
    } else if (arrLength == 1){
        return 0;
    }

    partition1(arr, 0, arrLength - 1);
    return 0;
}

int quickSort2(int *arr, int arrLength){
    if (arrLength <= 0){
        printf("Array length must be greater than 0");
        return -1;
    } else if (arrLength == 1){
        return 0;
    }

    partition2(arr, 0, arrLength - 1);
    return 0;
}

File: src/algorithms/sort/ts/quickSort.ts
export function quickSort1(values: number[]): void {
    if (values.length <= 1) {
        return
    }
    partition1(values, 0, values.length - 1)
}

export function quickSort2(values: number[]): void {
    if (values.length <= 1) {
        return
    }
    partition2(values, 0, values.length - 1)
}

function partition1<T>(values: T[], start: number, end: number): void {
    if (end <= start) {
        return
    }

    const pivotIndex = end
    let i = start

    for (let j = start; j < end; j++) {
        if (values[j] < values[pivotIndex]) {
            swap(values, i, j)
            i++
        }
    }
    swap(values, i, pivotIndex)
    partition1(values, start, i - 1)
    partition1(values, i + 1, end)
}

function partition2<T>(values: T[], start: number, end: number): void {
    if (end <= start) {
        return
    }

    const ref = values[start]
    let left = start + 1
    let right = end

    while (left <= right) {
        if (values[left] > ref && values[right] < ref) {
            swap(values, left, right)
        }

        if (values[left] <= ref) {
            left += 1
        }

        if (values[right] >= ref) {
            right -= 1
        }
    }

    const pivotIndex = right
    swap(values, pivotIndex, start)

    partition2(values, start, pivotIndex - 1)
    partition2(values, pivotIndex + 1, end)
}

function swap<T>(values: T[], i: number, j: number): void {
    if (i !== j) {
        const temp = values[j]
        values[j] = values[i]
        values[i] = temp
    }
}
